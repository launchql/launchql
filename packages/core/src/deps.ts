import { readFileSync } from 'fs';
import { sync as glob } from 'glob';
import { relative, join } from 'path';
import { parsePlanFile } from './files/plan/parser';
import { ExtendedPlanFile } from './files/types';
import { LaunchQLProject } from './class/launchql';

/**
 * Represents a dependency graph where keys are module identifiers
 * and values are arrays of their direct dependencies
 */
interface DependencyGraph {
  [key: string]: string[];
}

/**
 * Result object returned by dependency resolution functions
 */
interface DependencyResult {
  /** Array of external dependencies that are not part of the current project */
  external: string[];
  /** Array of modules in topologically sorted order for deployment */
  resolved: string[];
  /** The complete dependency graph mapping modules to their dependencies */
  deps: DependencyGraph;
}

/**
 * Configuration options for customizing the core dependency resolver behavior.
 * This interface allows different dependency resolution functions to share the same
 * core algorithm while maintaining their specific behaviors.
 */
interface DependencyResolverOptions {
  /** 
   * Function to handle external dependencies when they are not found in the dependency graph.
   * Used by extDeps to add external dependencies to the external array.
   * @param dep - The dependency module name that was not found
   * @param deps - The dependency graph to potentially add the dependency to
   * @param external - Array to collect external dependencies
   */
  handleExternalDep?: (dep: string, deps: DependencyGraph, external: string[]) => void;
  
  /** 
   * Function to transform module names and resolve their dependencies during resolution.
   * Used by getDeps and resolveDependencies to handle project:module syntax and tag resolution.
   * @param module - The module name to transform
   * @param extname - The current project/extension name for context
   * @returns Object containing the transformed module name, its dependency edges, and whether to return early
   */
  transformModule?: (module: string, extname: string) => { module: string; edges: string[] | undefined; returnEarly?: boolean };
  
  /** 
   * Function to generate standardized keys for dependency lookup in the graph.
   * Defaults to the module name as-is if not provided.
   * @param module - The module name to generate a key for
   * @returns The standardized key for dependency graph lookup
   */
  makeKey?: (module: string) => string;
  
  /** 
   * The current project/extension name used for internal reference resolution.
   * This is used to distinguish between internal and external module references.
   */
  extname: string;
}

/**
 * Core dependency resolution algorithm that handles circular dependency detection
 * and topological sorting of dependencies. This unified implementation eliminates
 * code duplication between getDeps, extDeps, and resolveDependencies.
 * 
 * @param deps - The dependency graph mapping modules to their dependencies
 * @param external - Array to collect external dependencies
 * @param options - Configuration options for customizing resolver behavior
 * @returns A function that performs dependency resolution with the given configuration
 */
function createDependencyResolver(
  deps: DependencyGraph,
  external: string[],
  options: DependencyResolverOptions
) {
  const { 
    handleExternalDep, 
    transformModule, 
    makeKey = (module: string) => module,
    extname 
  } = options;

  return function dep_resolve(
    sqlmodule: string,
    resolved: string[],
    unresolved: string[]
  ): void {
    unresolved.push(sqlmodule);
    
    let moduleToResolve = sqlmodule;
    let edges: string[] | undefined;
    let returnEarly = false;

    if (transformModule) {
      const result = transformModule(sqlmodule, extname);
      moduleToResolve = result.module;
      edges = result.edges;
      returnEarly = result.returnEarly || false;
    } else {
      edges = deps[makeKey(sqlmodule)];
    }

    // Handle external dependencies if no edges found
    if (!edges) {
      if (handleExternalDep) {
        handleExternalDep(sqlmodule, deps, external);
        edges = deps[sqlmodule] || [];
      } else {
        throw new Error(`Internal module not found: ${sqlmodule}`);
      }
    }

    if (returnEarly) {
      const index = unresolved.indexOf(sqlmodule);
      unresolved.splice(index, 1);
      return;
    }

    // Process each dependency
    for (const dep of edges) {
      if (!resolved.includes(dep)) {
        if (unresolved.includes(dep)) {
          throw new Error(`Circular reference detected ${moduleToResolve}, ${dep}`);
        }
        dep_resolve(dep, resolved, unresolved);
      }
    }

    resolved.push(moduleToResolve);
    const index = unresolved.indexOf(sqlmodule);
    unresolved.splice(index, 1);
  };
}

/**
 * Generates a standardized key for SQL deployment files
 * @param sqlmodule - The module name (e.g., 'users/create')
 * @returns The standardized file path key (e.g., '/deploy/users/create.sql')
 */
const makeKey = (sqlmodule: string): string => `/deploy/${sqlmodule}.sql`;

/**
 * Resolves dependencies for SQL deployment files in a package directory.
 * This is a simplified wrapper around resolveDependencies for backward compatibility.
 * 
 * @param packageDir - The package directory containing SQL files
 * @param extname - The extension/project name for internal reference resolution
 * @returns Object containing external dependencies, resolved order, and dependency graph
 */
export const getDeps = (
  packageDir: string,
  extname: string
): DependencyResult => {
  return resolveDependencies(packageDir, extname, { tagResolution: 'preserve' });
};

/**
 * Resolves dependencies for extension modules using a pre-built module map.
 * This is a simpler version that works with module metadata rather than parsing SQL files.
 * 
 * @param name - The name of the module to resolve dependencies for
 * @param modules - Record mapping module names to their dependency requirements
 * @returns Object containing external dependencies and resolved dependency order
 */
export const extDeps = (
  name: string,
  modules: Record<string, { requires: string[] }>
): { external: string[]; resolved: string[] } => {
  if (!modules[name]) {
    throw new Error(`module ${name} does not exist!`);
  }

  const external: string[] = [];
  const deps: DependencyGraph = Object.keys(modules).reduce((memo, key) => {
    memo[key] = modules[key].requires;
    return memo;
  }, {} as DependencyGraph);

  // Handle external dependencies for extDeps - simpler than getDeps
  const handleExternalDep = (dep: string, deps: DependencyGraph, external: string[]): void => {
    external.push(dep);
    deps[dep] = [];
  };

  // Create the dependency resolver with extDeps-specific configuration
  const dep_resolve = createDependencyResolver(deps, external, {
    handleExternalDep,
    extname: name // For extDeps, we use the module name as extname
  });

  const resolved: string[] = [];
  const unresolved: string[] = [];

  dep_resolve(name, resolved, unresolved);

  return { external, resolved };
};

export interface DependencyResolutionOptions {
  /**
   * How to handle tag references:
   * - 'preserve': Keep tags as-is, as changes in plan files
   * - 'resolve': Resolve tags to their target changes
   * - 'internal': Resolve internally but preserve in output
   */
  tagResolution?: 'preserve' | 'resolve' | 'internal';
  
  /**
   * Whether to load and parse plan files for tag resolution
   * Only applicable when tagResolution is 'resolve' or 'internal'
   */
  loadPlanFiles?: boolean;
  
  /**
   * Custom plan file loader function
   * Allows overriding the default plan file loading logic
   */
  planFileLoader?: (projectName: string, currentProject: string, packageDir: string) => ExtendedPlanFile | null;
}

/**
 * Unified dependency resolution function with configurable options.
 * This is the main entry point for all dependency resolution needs,
 * supporting advanced features like tag resolution and plan file loading.
 * 
 * @param packageDir - The package directory containing SQL files
 * @param extname - The extension/project name for internal reference resolution
 * @param options - Configuration options for dependency resolution behavior
 * @returns Object containing external dependencies, resolved order, and dependency graph
 */
export const resolveDependencies = (
  packageDir: string,
  extname: string,
  options: DependencyResolutionOptions = {}
): DependencyResult => {
  const {
    tagResolution = 'preserve',
    loadPlanFiles = true,
    planFileLoader
  } = options;
  
  // For 'resolve' and 'internal' modes, we need plan file loading
  const planCache: Record<string, ExtendedPlanFile> = {};
  
  // Helper function to load a plan file for a project
  const loadPlanFile = (projectName: string): ExtendedPlanFile | null => {
    if (!loadPlanFiles) {
      return null;
    }
    
    if (planFileLoader) {
      return planFileLoader(projectName, extname, packageDir);
    }
    
    if (planCache[projectName]) {
      return planCache[projectName];
    }
    
    try {
      let planPath: string;
      if (projectName === extname) {
        // For the current project
        planPath = join(packageDir, 'launchql.plan');
      } else {
        // For external projects, use LaunchQLProject to find module path
        const project = new LaunchQLProject(packageDir);
        const moduleMap = project.getModuleMap();
        const module = moduleMap[projectName];
        
        if (!module) {
          throw new Error(`Module ${projectName} not found in workspace`);
        }
        
        const workspacePath = project.getWorkspacePath();
        if (!workspacePath) {
          throw new Error(`No workspace found for module ${projectName}`);
        }
        
        planPath = join(workspacePath, module.path, 'launchql.plan');
      }
      
      const result = parsePlanFile(planPath);
      if (result.data) {
        planCache[projectName] = result.data;
        return result.data;
      }
    } catch (error) {
      // Plan file not found or parse error
      console.warn(`Could not load plan file for project ${projectName}: ${error}`);
    }
    
    return null;
  };
  
  // Helper function to resolve a tag to the change it represents
  const resolveTagToChange = (projectName: string, tagName: string): string | null => {
    const plan = loadPlanFile(projectName);
    if (!plan) {
      return null;
    }
    
    // Find the tag
    const tag = plan.tags.find(t => t.name === tagName);
    if (!tag) {
      return null;
    }
    
    // Return the change this tag points to
    return tag.change;
  };
  
  const external: string[] = [];
  const deps: DependencyGraph = {};
  // Keep track of tag mappings for internal resolution
  const tagMappings: Record<string, string> = {};

  // Process SQL files and build dependency graph
  const files = glob(`${packageDir}/deploy/**/*.sql`);

  for (const file of files) {
    const data = readFileSync(file, 'utf-8');
    const lines = data.split('\n');
    const key = '/' + relative(packageDir, file);
    deps[key] = [];

    for (const line of lines) {
      // Handle requires statements
      const requiresMatch = line.match(/^-- requires: (.*)/);
      if (requiresMatch) {
        const dep = requiresMatch[1].trim();
        
        // For 'preserve' mode, just add the dependency as-is (like original getDeps)
        if (tagResolution === 'preserve') {
          deps[key].push(dep);
          continue;
        }
        
        // For other modes, handle tag resolution
        if (dep.includes('@')) {
          const match = dep.match(/^([^:]+):@(.+)$/);
          if (match) {
            const [, projectName, tagName] = match;
            const taggedChange = resolveTagToChange(projectName, tagName);
            
            if (taggedChange) {
              if (tagResolution === 'resolve') {
                // Full resolution: replace tag with actual change
                const resolvedDep = `${projectName}:${taggedChange}`;
                deps[key].push(resolvedDep);
              } else if (tagResolution === 'internal') {
                // Internal resolution: keep tag in deps but track mapping
                tagMappings[dep] = `${projectName}:${taggedChange}`;
                deps[key].push(dep);
              }
            } else {
              // Could not resolve tag, keep it as is
              deps[key].push(dep);
            }
          } else {
            // Invalid tag format, keep as is
            deps[key].push(dep);
          }
        } else {
          // Not a tag, keep as is
          deps[key].push(dep);
        }
        continue;
      }

      // Handle deploy statements - exactly as in original
      let m2;
      let keyToTest;

      if (/:/.test(line)) {
        m2 = line.match(/^-- Deploy ([^:]*):([\w\/]+) to pg/);
        if (m2) {
          const actualProject = m2[1];
          keyToTest = m2[2];
        
          if (extname !== actualProject) {
            throw new Error(
              `Mismatched project name in deploy file:
          Expected project: ${extname}
          Found in line   : ${actualProject}
          Line            : ${line}`
            );
          }
        
          const expectedKey = makeKey(keyToTest);
          if (key !== expectedKey) {
            throw new Error(
              `Deployment script path or internal name mismatch:
          Expected key    : ${key}
          Found in line   : ${expectedKey}
          Line            : ${line}`
            );
          }
        }

      } else {
        m2 = line.match(/^-- Deploy (.*) to pg/);
        if (m2) {
          keyToTest = m2[1];
          if (key !== makeKey(keyToTest)) {
            throw new Error(
              'deployment script in wrong place or is named wrong internally\n' + line
            );
          }
        }
      }
    }
  }

  const transformModule = (sqlmodule: string, extname: string): { module: string; edges: string[] | undefined; returnEarly?: boolean } => {
    const originalModule = sqlmodule;
    
    // For 'preserve' mode, use simpler logic (like original getDeps)
    if (tagResolution === 'preserve') {
      let moduleToResolve = sqlmodule;
      let edges = deps[makeKey(sqlmodule)];
      
      if (/:/.test(sqlmodule)) {
        // Has a prefix — could be internal or external
        const [project, localKey] = sqlmodule.split(':', 2);
        if (project === extname) {
          // Internal reference to current project
          moduleToResolve = localKey;
          edges = deps[makeKey(localKey)];
          if (!edges) {
            throw new Error(`Internal module not found: ${localKey} (from ${project}:${localKey})`);
          }
        } else {
          // External reference — always OK, even if not in deps yet
          external.push(sqlmodule);
          deps[sqlmodule] = [];
          return { module: sqlmodule, edges: [], returnEarly: true };
        }
      } else {
        // No prefix — must be internal
        if (!edges) {
          throw new Error(`Internal module not found: ${sqlmodule}`);
        }
      }
      
      return { module: moduleToResolve, edges };
    }
    
    // Check if the ORIGINAL module (before tag resolution) is external
    if (/:/.test(originalModule)) {
      const [project, localKey] = originalModule.split(':', 2);
      if (project !== extname) {
        // External reference — always OK, even if not in deps yet
        external.push(originalModule);
        deps[originalModule] = deps[originalModule] || [];
        return { module: originalModule, edges: [], returnEarly: true };
      }
    }
    
    // For internal resolution mode, check if this module is a tag and resolve it
    let moduleToResolve = sqlmodule;
    if (tagResolution === 'internal' && tagMappings[sqlmodule]) {
      moduleToResolve = tagMappings[sqlmodule];
    }
    
    let edges = deps[makeKey(moduleToResolve)];
    
    if (/:/.test(moduleToResolve)) {
      // Has a prefix — must be internal since we already handled external above
      const [project, localKey] = moduleToResolve.split(':', 2);
      if (project === extname) {
        // Internal reference to current project
        moduleToResolve = localKey;
        edges = deps[makeKey(localKey)];
        if (!edges) {
          throw new Error(`Internal module not found: ${localKey} (from ${project}:${localKey})`);
        }
      }
    } else {
      // No prefix — must be internal
      if (!edges) {
        // Check if we have edges for the original module
        edges = deps[makeKey(sqlmodule)];
        if (!edges) {
          throw new Error(`Internal module not found: ${sqlmodule}`);
        }
      }
    }
    
    // For internal resolution, process dependencies through tag mappings
    if (tagResolution === 'internal' && edges) {
      const processedEdges = edges.map(dep => {
        // Check if this dependency is external - if so, don't resolve tags
        if (/:/.test(dep)) {
          const [project, localKey] = dep.split(':', 2);
          if (project !== extname) {
            // External dependency - keep original tag name
            return dep;
          }
        }
        
        // Internal dependency - apply tag mapping if available
        if (tagMappings[dep]) {
          return tagMappings[dep];
        }
        return dep;
      });
      return { module: moduleToResolve, edges: processedEdges };
    }
    
    return { module: moduleToResolve, edges };
  };

  // Create the dependency resolver with resolveDependencies-specific configuration
  const dep_resolve = createDependencyResolver(deps, external, {
    transformModule,
    makeKey,
    extname
  });

  let resolved: string[] = [];
  const unresolved: string[] = [];

  // Add synthetic root node - exactly as in original
  deps[makeKey('apps/index')] = Object.keys(deps)
    .filter((dep) => dep.startsWith('/deploy/'))
    .map((dep) => dep.replace(/^\/deploy\//, '').replace(/\.sql$/, ''));

  dep_resolve('apps/index', resolved, unresolved);
  
  // Remove synthetic root
  const index = resolved.indexOf('apps/index');
  resolved.splice(index, 1);
  delete deps[makeKey('apps/index')];

  // Sort extensions first - exactly as in original
  const extensions = resolved.filter((module) => module.startsWith('extensions/'));
  const normalSql = resolved.filter((module) => !module.startsWith('extensions/'));
  resolved = [...extensions, ...normalSql];

  return { external, resolved, deps };
};
