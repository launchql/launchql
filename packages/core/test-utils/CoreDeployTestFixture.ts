import { getEnvOptions } from '@launchql/env';
import { join } from 'path';
import { teardownPgPools } from 'pg-cache';
import { getPgEnvOptions } from 'pg-env';

import { LaunchQLProject } from '../src/core/class/launchql';
import { MigrateTestFixture, TestDatabase } from './index';
import { TestFixture } from './TestFixture';

export class CoreDeployTestFixture extends TestFixture {
  private databases: TestDatabase[] = [];
  private migrateFixture?: MigrateTestFixture;

  constructor(...fixturePath: string[]) {
    super(...fixturePath);
    this.migrateFixture = new MigrateTestFixture();
  }

  async setupTestDatabase(): Promise<TestDatabase> {
    if (!this.migrateFixture) {
      throw new Error('MigrateTestFixture not initialized');
    }
    
    const db = await this.migrateFixture.setupTestDatabase();
    this.databases.push(db);
    return db;
  }

  async deployModule(projectName: string, database: string, fixturePath: string[], logOnly: boolean = false): Promise<void> {
    const basePath = this.tempFixtureDir;
    const originalCwd = process.cwd();
    
    try {
      process.chdir(basePath);
      
      const project = new LaunchQLProject(basePath);
      
      const opts = getEnvOptions({ 
        pg: getPgEnvOptions({ database }),
        deployment: {
          fast: false,
          usePlan: true,
          logOnly
        }
      });

      await project.deploy(opts, projectName, undefined, true);
    } finally {
      process.chdir(originalCwd);
    }
  }

  async revertModule(projectName: string, database: string, fixturePath: string[], toChange?: string): Promise<void> {
    const basePath = this.tempFixtureDir;
    const originalCwd = process.cwd();
    
    try {
      const projectPath = join(basePath, 'packages', projectName);
      process.chdir(projectPath);
      
      const project = new LaunchQLProject(projectPath);
      
      const opts = getEnvOptions({ 
        pg: getPgEnvOptions({ database })
      });

      await project.revert(opts, projectName, toChange, true);
    } finally {
      process.chdir(originalCwd);
    }
  }

  async verifyModule(projectName: string, database: string, fixturePath: string[], toChange?: string): Promise<void> {
    const basePath = this.tempFixtureDir;
    const originalCwd = process.cwd();
    
    try {
      process.chdir(basePath);
      
      const project = new LaunchQLProject(basePath);
      
      const opts = getEnvOptions({ 
        pg: getPgEnvOptions({ database })
      });

      await project.verify(opts, projectName, toChange, true);
    } finally {
      process.chdir(originalCwd);
    }
  }

  async cleanup(): Promise<void> {
    for (const db of this.databases) {
      await db.close();
    }
    this.databases = [];
    
    if (this.migrateFixture) {
      await this.migrateFixture.cleanup();
    }
    
    await teardownPgPools();
    
    // Small delay to ensure connections are fully closed
    await new Promise(resolve => setTimeout(resolve, 10));
    
    super.cleanup();
  }
}
