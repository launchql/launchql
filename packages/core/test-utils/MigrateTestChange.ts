export interface MigrateTestChange {
  name: string;
  dependencies?: string[];
  timestamp?: string;
  planner?: string;
  email?: string;
  comment?: string;
}
