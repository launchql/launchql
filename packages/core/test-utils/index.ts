export { TestPlan } from './TestPlan';
export { CoreDeployTestFixture } from './CoreDeployTestFixture';
export { TestFixture } from './TestFixture';

export { MigrateTestFixture } from './MigrateTestFixture';
export { TestDatabase } from './TestDatabase';
export { MigrateTestChange } from './MigrateTestChange';

export {
  FIXTURES_PATH,
  getFixturePath,
  cleanText,
  teardownAllPools,
  closeDatabasePools
} from './utils';
