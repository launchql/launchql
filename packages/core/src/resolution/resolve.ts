import { readFileSync } from 'fs';

import { getChanges, getExtensionName } from '../files';
import { parsePlanFile } from '../files/plan/parser';
import { resolveDependencies } from './deps';

/**
 * Resolves SQL scripts for deployment or reversion.
 *
 * @param pkgDir - The package directory (defaults to the current working directory).
 * @param scriptType - The type of script to resolve (`deploy` or `revert`).
 * @returns A single concatenated SQL script as a string.
 */
export const resolve = (
  pkgDir: string = process.cwd(),
  scriptType: 'deploy' | 'revert' = 'deploy'
): string => {
  const sqlfile: string[] = [];
  const name = getExtensionName(pkgDir);
  const { resolved, external } = resolveDependencies(pkgDir, name, { tagResolution: 'resolve' });

  const scripts = scriptType === 'revert' ? [...resolved].reverse() : resolved;

  for (const script of scripts) {
    if (external.includes(script)) continue;
    const file = `${pkgDir}/${scriptType}/${script}.sql`;
    const dscript = readFileSync(file, 'utf-8');
    sqlfile.push(dscript);
  }

  return sqlfile.join('\n');
};

/**
 * Resolves SQL scripts based on the `launchql.plan` file.
 *
 * @param pkgDir - The package directory (defaults to the current working directory).
 * @param scriptType - The type of script to resolve (`deploy` or `revert`).
 * @returns A single concatenated SQL script as a string.
 */
export const resolveWithPlan = (
  pkgDir: string = process.cwd(),
  scriptType: 'deploy' | 'revert' = 'deploy'
): string => {
  const sqlfile: string[] = [];
  const planPath = `${pkgDir}/launchql.plan`;
  
  let resolved = getChanges(planPath);

  if (scriptType === 'revert') {
    resolved = resolved.reverse();
  }

  for (const script of resolved) {
    const file = `${pkgDir}/${scriptType}/${script}.sql`;
    const dscript = readFileSync(file, 'utf-8');
    sqlfile.push(dscript);
  }

  return sqlfile.join('\n');
};

/**
 * Resolves a tag reference to its corresponding change name.
 * Tags provide a way to reference specific points in a package's deployment history.
 * 
 * @param planPath - Path to the plan file containing tag definitions
 * @param tagReference - The tag reference to resolve (e.g., "package:@tagName" or "@tagName")
 * @param currentPackage - The current package name (used when tag doesn't specify package)
 * @returns The resolved change name
 * @throws Error if tag format is invalid or tag is not found
 * 
 * @example
 * // Resolve a tag in the current package
 * resolveTagToChangeName('/path/to/launchql.plan', '@v1.0.0', 'mypackage')
 * // Returns: 'schema/v1'
 * 
 * @example
 * // Resolve a tag from another package
 * resolveTagToChangeName('/path/to/launchql.plan', 'auth:@v2.0.0')
 * // Returns: 'users/table'
 */
export const resolveTagToChangeName = (
  planPath: string, 
  tagReference: string, 
  currentProject?: string
): string => {
  // If not a tag reference, return as-is
  if (!tagReference.includes('@')) {
    return tagReference;
  }
  
  // Handle simple tag format (@tagName) by prepending current package
  if (tagReference.startsWith('@') && !tagReference.includes(':')) {
    if (!currentProject) {
      const plan = parsePlanFile(planPath);
      if (!plan.data) {
        throw new Error(`Could not parse plan file: ${planPath}`);
      }
      currentProject = plan.data.package;
    }
    tagReference = `${currentProject}:${tagReference}`;
  }
  
  // Parse package:@tagName format
  const match = tagReference.match(/^([^:]+):@(.+)$/);
  if (!match) {
    throw new Error(`Invalid tag format: ${tagReference}. Expected format: package:@tagName or @tagName`);
  }
  
  const [, projectName, tagName] = match;
  
  // Parse plan file to find tag
  const planResult = parsePlanFile(planPath);
  
  if (!planResult.data) {
    throw new Error(`Could not parse plan file: ${planPath}`);
  }
  
  // Find the tag in the plan
  const tag = planResult.data.tags?.find((t: any) => t.name === tagName);
  if (!tag) {
    throw new Error(`Tag ${tagName} not found in project ${projectName}`);
  }
  
  return tag.change;
};
