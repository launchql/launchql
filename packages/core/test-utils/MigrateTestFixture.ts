import { cpSync, mkdirSync,mkdtempSync, rmSync, writeFileSync } from 'fs';
import { tmpdir } from 'os';
import { join } from 'path';
import { Pool } from 'pg';
import { getPgPool } from 'pg-cache';
import { PgConfig } from 'pg-env';
import { getPgEnvOptions } from 'pg-env';

import { LaunchQLMigrate } from '../src/migrate/client';
import { MigrateTestChange } from './MigrateTestChange';
import { TestDatabase } from './TestDatabase';
import { FIXTURES_PATH } from './utils';

export class MigrateTestFixture {
  private tempDirs: string[] = [];
  private databases: TestDatabase[] = [];
  private dbCounter = 0;
  private pools: Pool[] = [];

  async setupTestDatabase(): Promise<TestDatabase> {
    const dbName = `test_migrate_${Date.now()}_${Math.random().toString(36).substring(2, 8)}_${this.dbCounter++}`;
    
    // Get base config from environment using pg-env
    const baseConfig = getPgEnvOptions({
      database: 'postgres'
    });
    
    // Create database using admin pool
    const adminPool = getPgPool(baseConfig);
    try {
      await adminPool.query(`CREATE DATABASE "${dbName}"`);
    } catch (e) {
      if (e && typeof e === 'object' && 'errors' in e && Array.isArray((e as any).errors)) {
        for (const err of (e as any).errors) {
          console.error('AggregateError item:', err);
        }
      } else {
        console.error('Test failure:', e);
      }
      throw e;
    }
    
    // Get config for the new test database
    const pgConfig = getPgEnvOptions({
      database: dbName
    });
    
    const config: PgConfig = {
      host: pgConfig.host,
      port: pgConfig.port,
      user: pgConfig.user,
      password: pgConfig.password,
      database: pgConfig.database
    };

    // Initialize migrate schema
    const migrate = new LaunchQLMigrate(config);
    await migrate.initialize();

    // Get pool for test database operations
    const pool = getPgPool(pgConfig);
    this.pools.push(pool);

    const fixture = this;
    const db: TestDatabase = {
      name: dbName,
      config,
      
      async query(sql: string, params?: any[]) {
        return pool.query(sql, params);
      },

      async exists(type: 'schema' | 'table', name: string) {
        if (type === 'schema') {
          const result = await pool.query(
            `SELECT EXISTS (
              SELECT 1 FROM information_schema.schemata 
              WHERE schema_name = $1
            ) as exists`,
            [name]
          );
          return result.rows[0].exists;
        } else {
          const [schema, table] = name.includes('.') ? name.split('.') : ['public', name];
          const result = await pool.query(
            `SELECT EXISTS (
              SELECT 1 FROM information_schema.tables 
              WHERE table_schema = $1 AND table_name = $2
            ) as exists`,
            [schema, table]
          );
          return result.rows[0].exists;
        }
      },

      async getDeployedChanges() {
        const result = await pool.query(
          `SELECT project, change_name, deployed_at 
           FROM launchql_migrate.changes 
           ORDER BY deployed_at`
        );
        return result.rows;
      },

      async getMigrationState() {
        const changes = await pool.query(`
          SELECT project, change_name, script_hash, deployed_at
          FROM launchql_migrate.changes 
          ORDER BY deployed_at
        `);
        
        const events = await pool.query(`
          SELECT project, change_name, event_type, occurred_at, error_message, error_code
          FROM launchql_migrate.events 
          ORDER BY occurred_at
        `);
        
        const sanitizedEvents = events.rows;
        
        // Remove timestamps from objects for consistent snapshots
        const cleanChanges = changes.rows.map(({ deployed_at, ...change }) => change);
        const cleanEvents = sanitizedEvents.map(({ occurred_at, ...event }) => event);
        
        return {
          changes: cleanChanges,
          events: cleanEvents,
          changeCount: cleanChanges.length,
          eventCount: cleanEvents.length
        };
      },

      async getDependencies(project: string, changeName: string) {
        const result = await pool.query(
          `SELECT d.requires 
           FROM launchql_migrate.dependencies d
           JOIN launchql_migrate.changes c ON c.change_id = d.change_id
           WHERE c.project = $1 AND c.change_name = $2`,
          [project, changeName]
        );
        return result.rows.map((row: any) => row.requires);
      },

      async close() {
        // Don't close the pool here as it's managed by pg-cache
        // Just mark this database as closed
      }
    };

    this.databases.push(db);
    return db;
  }


  setupFixture(fixturePath: string[]): string {
    const originalPath = join(FIXTURES_PATH, ...fixturePath);
    const fixtureName = fixturePath[fixturePath.length - 1]; // Use last element as fixture name
    const tempDir = mkdtempSync(join(tmpdir(), 'migrate-test-'));
    const fixtureDestPath = join(tempDir, fixtureName);
    
    cpSync(originalPath, fixtureDestPath, { recursive: true });
    this.tempDirs.push(tempDir);
    
    return fixtureDestPath;
  }

  createPlanFile(project: string, changes: MigrateTestChange[]): string {
    const tempDir = mkdtempSync(join(tmpdir(), 'migrate-test-'));
    this.tempDirs.push(tempDir);

    const lines = [
      '%syntax-version=1.0.0',
      `%project=${project}`,
      `%uri=https://github.com/test/${project}`,
      ''
    ];

    for (const change of changes) {
      let line = change.name;
      
      if (change.dependencies && change.dependencies.length > 0) {
        line += ` [${change.dependencies.join(' ')}]`;
      }
      
      line += ` ${change.timestamp || new Date().toISOString().replace(/\.\d{3}Z$/, 'Z')}`;
      line += ` ${change.planner || 'test'}`;
      line += ` <${change.email || 'test@example.com'}>`;
      
      if (change.comment) {
        line += ` # ${change.comment}`;
      }
      
      lines.push(line);
    }

    const planPath = join(tempDir, 'launchql.plan');
    writeFileSync(planPath, lines.join('\n'));
    
    return tempDir;
  }

  createScript(dir: string, type: 'deploy' | 'revert' | 'verify', name: string, content: string): void {
    const scriptDir = join(dir, type);
    mkdirSync(scriptDir, { recursive: true });
    writeFileSync(join(scriptDir, `${name}.sql`), content);
  }

  async cleanup(): Promise<void> {
    // Close all test database pools FIRST
    for (const pool of this.pools) {
      try {
        await pool.end();
      } catch (e) {
        // Ignore errors during pool closure
      }
    }
    
    // Clear the pools array
    this.pools = [];

    // Remove temporary directories
    for (const dir of this.tempDirs) {
      try {
        rmSync(dir, { recursive: true, force: true });
      } catch (e) {
        // Ignore errors during cleanup
      }
    }
    
    // Clear the databases array
    this.databases = [];
  }
}
