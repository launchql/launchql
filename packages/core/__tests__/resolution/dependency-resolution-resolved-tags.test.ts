import { TestFixture } from '../../test-utils';
import { resolveDependencies } from '../../src/resolution/deps';

let fixture: TestFixture;

beforeAll(() => {
  fixture = new TestFixture('sqitch');
});

afterAll(() => {
  fixture.cleanup();
});

it('sqitch package dependencies with resolved tags [simple-w-tags/1st]', async () => {
  const res = await resolveDependencies(
    fixture.getFixturePath('simple-w-tags', 'packages', 'my-first'),
    'my-first',
    { tagResolution: 'resolve' }
  );
  expect(res).toMatchSnapshot();
});

it('sqitch package dependencies with resolved tags [simple-w-tags/2nd]', async () => {
  const res = await resolveDependencies(
    fixture.getFixturePath('simple-w-tags', 'packages', 'my-second'),
    'my-second',
    { tagResolution: 'resolve' }
  );
  expect(res).toMatchSnapshot();
});

it('sqitch package dependencies with resolved tags [simple-w-tags/3rd]', async () => {
  const res = await resolveDependencies(
    fixture.getFixturePath('simple-w-tags', 'packages', 'my-third'),
    'my-third',
    { tagResolution: 'resolve' }
  );
  expect(res).toMatchSnapshot();
});
