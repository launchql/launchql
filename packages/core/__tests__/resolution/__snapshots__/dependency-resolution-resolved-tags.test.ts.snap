// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sqitch package dependencies with resolved tags [simple-w-tags/1st] 1`] = `
{
  "deps": {
    "/deploy/schema_myapp.sql": [],
    "/deploy/table_products.sql": [
      "my-first:schema_myapp",
      "my-first:table_users",
    ],
    "/deploy/table_users.sql": [
      "my-first:schema_myapp",
    ],
  },
  "external": [],
  "resolved": [
    "schema_myapp",
    "table_users",
    "schema_myapp",
    "schema_myapp",
    "table_users",
    "table_products",
  ],
  "resolvedTags": {},
}
`;

exports[`sqitch package dependencies with resolved tags [simple-w-tags/2nd] 1`] = `
{
  "deps": {
    "/deploy/create_another_table.sql": [
      "my-second:create_table",
    ],
    "/deploy/create_schema.sql": [
      "my-first:table_users",
    ],
    "/deploy/create_table.sql": [
      "my-second:create_schema",
    ],
    "my-first:table_users": [],
  },
  "external": [
    "my-first:table_users",
    "my-first:table_users",
  ],
  "resolved": [
    "create_schema",
    "create_table",
    "create_schema",
    "create_table",
    "create_another_table",
  ],
  "resolvedTags": {},
}
`;

exports[`sqitch package dependencies with resolved tags [simple-w-tags/3rd] 1`] = `
{
  "deps": {
    "/deploy/create_schema.sql": [
      "my-first:table_products",
      "my-second:create_table",
    ],
    "/deploy/create_table.sql": [
      "my-third:create_schema",
      "my-second:create_another_table",
    ],
    "my-first:table_products": [],
    "my-second:create_another_table": [],
    "my-second:create_table": [],
  },
  "external": [
    "my-first:table_products",
    "my-second:create_table",
    "my-second:create_another_table",
  ],
  "resolved": [
    "create_schema",
    "create_table",
  ],
  "resolvedTags": {},
}
`;
